design/batch_retrieval_design.md

```diff
@@ -1,20 +1,352 @@
+
+```json
+{
+  "variables": [
+    {"id": "var/path/conjurSecret"},
+    {"id": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+#### Headers
+
+```
+Content-Type: application/json
+Authorization: Token token="eyJkYX...Rhb="
+```
+
+### Response
+
+#### Batch Response
+
+```json
+# Headers
+HTTP/1.1
+Content-Type: application/json
+
+# Status response code
+200 OK
+
+# Body
+{
+  "variables": [
+  {
+    "id": "var/path/name1",
+    "status": 200,
+    "value": "superSecret"
+  },
+  {
+    "id": "var/path/name2",
+    "status": 403,
+    "error": {
+      "message": "Forbidden"
+    }
+  },
+  {
+    "id": "var/path/name3",
+    "status": 404,
+    "error": {
+      "message": "Not Found"
+ 		}
+  },
+  {
+    "id": "var/path/name4",
+    "status": 200,
+    "value": "superSuperSecret"    
+  }
+ ]
+}
+```
+
+| Code            | Description                                                  |
+| --------------- | ------------------------------------------------------------ |
+| 200 OK          | All secret values were retrieved successfully                |
+| 400 Bad Request | Server cannot process request due to client error (unparsable request body) |
+
+If no Conjur/DAP secrets were able to be retrieved, the response will still return a 200 OK because the [request body is parsable]([https://docs.microsoft.com/en-us/graph/json-batching#:~:text=The%20status%20code%20on%20a,requests%20inside%20the%20batch%20succeeded](https://docs.microsoft.com/en-us/graph/json-batching#:~:text=The status code on a,requests inside the batch succeeded)) and request was handled.
+
+If an invalid request body was supplied, a 400 Bad Request will be returned.
+
+#### Response body
+
+All secret values will be treated as individual entities. If a single secret fails to be retrieved, this will ***not*** impact the integrity of the request. Each secret has it's own response code. 
+
+| Parameter     | Description                                        |
+| ------------- | -------------------------------------------------- |
+| id            | ID for Conjur/DAP variable                         |
+| value         | Secret for variable ID                             |
+| error.status  | Response status code of the error. See table below |
+| error.message | Response message of the error. See table below     |
+
+##### Individual secret responses
+
+The following are the possible error codes that can be returned for individual secrets.
+
+| Status/Message   | Description                                                  |
+| ---------------- | ------------------------------------------------------------ |
+| 400 Bad Request  | Server cannot process request due to client error (invalid input parameters or logical validations). For example: adding `<>` in path to secret in Conjur/DAP |
+| 401 Unauthorized | The request lacks valid authentication credentials           |
```

Comment:
`401` is a result of invalid Access Token.
Access Token validation occurs once even before flow reaches a controller code (please fix me if I'm wrong).
So `401` is a return code of the whole API rather than an "Individual secret response".

Reply:
You are correct. Good catch